avm/src/vm/opcode.rs

```diff
@@ -1560,90 +1556,80 @@ pub static OPCODES: Lazy<HashMap<i64, ByteOpcode>> = Lazy::new(|| {
+      hand_mem.push_register_from_fractal(args[2], &record, 0);
     } else {
       hand_mem.push_fixed(args[2], 0i64);
       let error_string = last.err().unwrap();
-      hand_mem.push_fractal(args[2], &HandlerMemory::str_to_fractal(&error_string));
+      hand_mem.push_fractal(args[2], HandlerMemory::str_to_fractal(&error_string));
     }
     None
   });
   cpu!("delindx", |args, hand_mem| {
     let idx = hand_mem.read_fixed(args[1]) as usize;
     let el = hand_mem.delete(args[0], idx);
-    hand_mem.write_fractal(args[2], &Vec::new());
+    hand_mem.init_fractal(args[2]);
     if el.is_ok() {
       hand_mem.push_fixed(args[2], 1i64);
       let record = el.ok().unwrap();
-      let (a, b) = record;
-      hand_mem.push_idxs(args[2], a, b as usize);
+      hand_mem.push_register_from_fractal(args[2], &record, 0);
     } else {
       hand_mem.push_fixed(args[2], 0i64);
       let error_string = el.err().unwrap();
-      hand_mem.push_fractal(args[2], &HandlerMemory::str_to_fractal(&error_string));
+      hand_mem.push_fractal(args[2], HandlerMemory::str_to_fractal(&error_string));
     }
     None
   });
   cpu!("newarr", |args, hand_mem| {
-    hand_mem.write_fractal(args[2], &Vec::new());
+    hand_mem.init_fractal(args[2]);
     None
   });
   io!("map", |args, mut hand_mem| {
     Box::pin(async move {
-      let arr = hand_mem.read_fractal(args[0]);
-      let len = arr.len();
+      let fractal = hand_mem.read_fractal(args[0]);
       let subhandler = HandlerFragment::new(args[1], 0);
       let mut mappers = Vec::new();
-      for i in 0..len {
+      for i in 0..fractal.len() {
         let mut hm = hand_mem.fork();
-        hm.set_addr(CLOSURE_ARG_MEM_START + 1, arr[i].0, arr[i].1 as usize);
+        hm.register_from_fractal(CLOSURE_ARG_MEM_START + 1, &fractal, i);
         hm.write_fixed(CLOSURE_ARG_MEM_START + 2, i as i64);
         mappers.push(subhandler.clone().run(hm));
       }
       let hms = join_all(mappers).await;
-      hand_mem.write_fractal(args[2], &Vec::new());
+      hand_mem.init_fractal(args[2]);
       for hm in hms {
         hand_mem.join(hm);
-        let (a, b) = hand_mem.addr_to_idxs(CLOSURE_ARG_MEM_START);
-        hand_mem.push_idxs(args[2], a, b);
+        hand_mem.push_register(args[2], CLOSURE_ARG_MEM_START);
       }
       hand_mem
     })
   });
   io!("mapl", |args, mut hand_mem| {
     Box::pin(async move {
-      let arr = hand_mem.read_fractal(args[0]);
-      let arrv = arr.to_vec();
-      let len = arrv.len();
-      drop(arr);
+      let fractal = hand_mem.read_fractal(args[0]);
       let subhandler = HandlerFragment::new(args[1], 0);
-      hand_mem.write_fractal(args[2], &Vec::new());
-      for i in 0..len {
-        hand_mem.set_addr(CLOSURE_ARG_MEM_START + 1, arrv[i].0, arrv[i].1 as usize);
+      hand_mem.init_fractal(args[2]);
+      for i in 0..fractal.len() {
+        hand_mem.register_from_fractal(CLOSURE_ARG_MEM_START + 1, &fractal, i);
```

Comment:
I don't think you need this. Just `hand_mem.register_out(args[0], i, CLOSURE_ARG_MEM_START + 1)`

Reply:
If it's only used in situations like this, we can probably drop the `register_from_fractal` method.

Reply:
hmm I just realized that might be true, but they seem pretty equivalent. Your suggestion happens to work because of the way to we write into the closure argument memory space. The code that uses `register_from_fractal` to make more sense since we are iterating over the length of fractal? Your suggestion of using `register_out ` seems a bit more "magical" to follow to me or not a clear improvement at least?

Reply:
what? But `register_out` was explicitly designed to tackle this?

* `register` -> make an alias from address A to address B
* `register_in` -> make an alias from address A to index X of address B
* `register_out` -> make an alias from index X of address A to address B

Reply:
I see it as less magical than a `register_from_fractal` that's gonna have to do some crazy magic once you add the HandlerMemory inheritance functionality to get the perf improvement we're after -- since that version is gonna have to figure out if the Fractal it was given is actually from a nested HandlerMemory and then copy it to the new HandlerMemory.

Reply:
> I see it as less magical than a `register_from_fractal` that's gonna have to do some crazy magic once you add the HandlerMemory inheritance functionality to get the perf improvement we're after -- since that version is gonna have to figure out if the Fractal it was given is actually from a nested HandlerMemory and then copy it to the new HandlerMemory.

now that is a very good point that I hadn't factored in!